File: src/main/java/com/braydenoneal/blang/parser/expression/ExpressionType.java
package com.braydenoneal.blang.parser.expression;

import com.mojang.serialization.Lifecycle;
import com.mojang.serialization.MapCodec;
import net.minecraft.registry.Registry;
import net.minecraft.registry.RegistryKey;
import net.minecraft.registry.SimpleRegistry;
import net.minecraft.util.Identifier;

public record ExpressionType<T extends Expression>(MapCodec<T> codec) {
    public static final Registry<ExpressionType<?>> REGISTRY = new SimpleRegistry<>(
            RegistryKey.ofRegistry(Identifier.of("blogic", "expression_types")), Lifecycle.stable()
    );
}

File: src/main/kotlin/com/braydenoneal/blang/parser/expression/ExpressionType.kt
package com.braydenoneal.blang.parser.expression

import com.mojang.serialization.Lifecycle
import com.mojang.serialization.MapCodec
import net.minecraft.registry.Registry
import net.minecraft.registry.RegistryKey
import net.minecraft.registry.SimpleRegistry
import net.minecraft.util.Identifier

data class ExpressionType<T : Expression>(val codec: MapCodec<T>) {
    companion object {
        val REGISTRY: Registry<ExpressionType<*>> = SimpleRegistry(
            RegistryKey.ofRegistry(Identifier.of("blogic", "expression_types")), Lifecycle.stable()
        )
    }
}
